sdk/communication/Azure.Communication.CallingServer/src/ContentDownloader.cs

```diff
@@ -0,0 +1,142 @@
+                _client._pipeline.ResponseClassifier,
+                Constants.ContentDownloader.RetriableStreamRetries
+                );
+
+            return Response.FromValue(stream, response.GetRawResponse());
+        }
+
+        internal async Task<Response> StagedDownloadAsync(Stream destination, Uri endpoint,
+            ContentTransferOptions transferOptions = default, bool async = true, CancellationToken cancellationToken = default)
+        {
+            PartitionedDownloader downloader = new(_client, transferOptions);
+            if (async)
+            {
+                return await downloader.DownloadToAsync(destination, endpoint, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                return downloader.DownloadTo(destination, endpoint, cancellationToken);
+            }
+        }
+
+        private async Task<Response<Stream>> StartDownloadAsync(Uri endpoint, HttpRange range = default, long startOffset = 0,
+            bool async = true, CancellationToken cancellationToken = default)
+        {
+            HttpRange? pageRange = null;
+
+            if (range != default || startOffset != 0)
+            {
+                pageRange = new HttpRange(
+                    range.Offset + startOffset,
+                    range.Length.HasValue ?
+                        range.Length.Value - startOffset :
+                        null);
+            }
+
+            HttpMessage message = GetHttpMessage(endpoint, pageRange);
+            if (async)
+            {
+                await _client._pipeline.SendAsync(message, cancellationToken).ConfigureAwait(false);
+                switch (message.Response.Status)
+                {
+                    case 200:
+                    case 206:
+                        {
+                            Stream value = message.ExtractResponseContent();
+                            return Response.FromValue(value, message.Response);
+                        }
+                    default:
+                        throw await _client._clientDiagnostics.CreateRequestFailedExceptionAsync(message.Response).ConfigureAwait(false);
+                }
+            }
+            else
+            {
+                _client._pipeline.Send(message, cancellationToken);
+
+                switch (message.Response.Status)
+                {
+                    case 200:
+                    case 206:
+                        {
+                            Stream value = message.ExtractResponseContent();
+                            return Response.FromValue(value, message.Response);
+                        }
+                    default:
+                        throw _client._clientDiagnostics.CreateRequestFailedException(message.Response);
+                }
+            }
+        }
+
+        private HttpMessage GetHttpMessage(Uri endpoint, HttpRange? rangeHeader = null)
```

Comment:
Synced offline on this:
- Franko: _The biggest reason why I went this route was basically because I thought that Autorest was autogenerated using the Swagger. Since we don't have a REST endpoint, hence no Swagger entry, I thought that Autorest was not possible._
_If that is not true, how can I create such Autorest methods?_
- Dominik: _Hm, that's a new scenario. Could you write a swagger for the service you're using regardless and use the hand-written swagger to generate code?_
- Franko: _I could probably investigate on that. The issue I have is the timeline we have for our Public preview launch (Next monday).
Do you think that I can do this investigation and if the approach is feasible the changes before the GA launch?_
- Dominik: _Yes, that sounds good to me_ 👍